[SR-1738] Allow *only* static libraries to be built

modocache opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?
<!-- Description about pull request. -->

This splits the `--build-swift-stdlib` and `--build-swift-sdk-overlay` arguments into `dynamic` and `static` variants, which makes the following build command possible:

```
utils/build-script -- \
  --build-swift-dynamic-stdlib=0 --build-swift-dynamic-sdk-overlay=0 \
  --build-swift-static-stdlib=1 --build-swift-static-sdk-overlay=0
```

This command produces *only* static libraries for the stdlib, and no SDK overlay libraries at all. Many other finely-grained build options are now possible.

@gribozavr This pull request is built on top of https://github.com/apple/swift/pull/3020 and https://github.com/apple/swift/pull/3043; it'll be easier to review and merge those first. Thanks for all your feedback on this! 🙇 

#### Resolved bug number: ([SR-1738](https://bugs.swift.org/browse/SR-1738))
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        All supported platforms     | @swift-ci Please smoke test and merge
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

**Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        All supported platforms     | @swift-ci Please test and merge
        OS X platform               | @swift-ci Please test OS X platform
        OS X platform               | @swift-ci Please benchmark
        Linux platform              | @swift-ci Please test Linux platform


**Lint Testing**

        Language     | Comment
        ------------ | -------------
        Python       | @swift-ci Please Python lint

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

modocache commented:
@swift-ci please test

modocache commented:
@swift-ci please test

modocache commented:
@gribozavr This was a sort of work-in-progress, but I've updated it to something I think is good to go. It's based on the work from #3020 and #3043, so it'd be easier to review/merge those before looking at this. Let me know what you think -- thanks!

gribozavr commented:
Thank you!  Would you mind rebasing?

modocache commented:
Thanks @gribozavr. It's weird, though, this rebased cleanly onto master. I wonder if GitHub determines conflicts differently somehow... oh well!

@swift-ci please test

gribozavr commented on utils/build-script-impl:
```diff
@@ -154,10 +154,11 @@ KNOWN_SETTINGS=(
-    build-swift-stdlib          "1"              "set to 1 to build the Swift standard library"
-    build-swift-stdlib-unittest-extra "0"        "set to 1 to build optional StdlibUnittest components"
-    build-swift-sdk-overlay     "1"              "set to 1 to build the Swift SDK overlay"
+    build-swift-dynamic-stdlib  "1"              "set to 1 to build dynamic variants of the Swift standard library"
```
Could we keep `build-swift-stdlib` and `build-swift-sdk-overlay` as aliases for `--build-swift-static-xyz=1 --build-swift-dynamic-xyz=1`?  We are using these in build presets, would be best to not break them.

modocache commented:
Looks like the OS X build has stalled for the past 7 hours. Kicking it.

@swift-ci please test

modocache commented:
@gribozavr OK! I think this is good to go.

gribozavr commented on utils/build-script-impl:
```diff
@@ -810,6 +813,14 @@ if [[ "${CROSS_COMPILE_HOSTS}" ]]; then
 fi
 
+if [[ "${BUILD_SWIFT_STDLIB}" ]]; then
+    BUILD_SWIFT_DYNAMIC_STDLIB=1
```
Shouldn't it turn off the static one as well?

Also, would you mind moving the alias processing into Python?  (Possibly in a future change.)

gribozavr replied:
Could you add support for `build-swift-sdk-overlay=0`?  Some of our internal presets use it.

gribozavr commented:
Thank you!  I'm running a few tests and I'll merge soon!

gribozavr commented:
@modocache Sorry, compatibility with `build-swift-sdk-overlay=0` is a blocker for us.  Would you mind taking a look?

modocache commented:
@gribozavr OK, I've addressed your comments on both moving the logic into Python and on `build-swift-sdk-overlay` 👍

modocache commented:
@swift-ci please test

modocache commented:
@gribozavr Is there a way to test the compatibility concerns you mentioned? @swift-ci consistently passes for me. Could you confirm whether this passes, and whether it can be merged? Sorry for all the pings 😅

gribozavr commented:
Sorry, I was reviewing other patches.  I'm running a test locally now.

gribozavr commented:
@modocache Sorry, this change does not yet fix the issue.  The command that we care about is:

```
./utils/build-script -R --build-swift-sdk-overlay=0
```

It should not build the overlay at all.

gribozavr commented on utils/build-script:
```diff
@@ -353,6 +353,21 @@ class BuildScriptInvocation(object):
+        # options.
+        if args.build_swift_stdlib:
+            args.build_swift_dynamic_stdlib = True
+            args.build_swift_static_stdlib = False
```
Could you add an `else` clause that sets both to `False`?

modocache replied:
Wouldn't that overwrite whatever was passed to `--build-swift-dynamic-stdlib` and `--build-swift-static-stdlib`? For example, consider the following invocation:

```
utils/build-script --build-swift-static-stdlib=1
```

Since `--build-swift-stdlib` is not passed in, the `else` statement would be executed, which would set `--build-swift-static-stdlib=0`, thus negating the desired invocation. Or am I missing something?

gribozavr replied:
You're right.  We probably want it to be a tristate then.  If `--build-swift-stdlib` is not passed on the command line, do nothing, if it is passed, then it should override other args.

Once this PR is merged, I'll try to remove the uses of legacy parameters internally.

modocache replied:
Is there a good way to determine whether `--build-swift-stdlib` is specified on the command line? Right now it seems like the following are indistinguishable:

```
$ utils/build-script
$ utils/build-script --build-swift-stdlib=0
```

Is there a good way to distinguish these two using the existing `swift_build_support` code? Or should I make some modifications there, or just inspect `argv`?

Thanks for all your help!

gribozavr replied:
There's a way, but we'd need to stop using the `optional_bool` action.  Since this is a short-lived compatibility aid, I think inspecting `argv` is acceptable.

rintaro replied:
I think `action=optional_bool, default=None` will have 3 possible state.
(`is None`, `True`, `False`).
Not confirmed though.

modocache replied:
Perfect, this is just what I need. Thanks, @rintaro!! I'll make the changes first thing tomorrow.

gribozavr commented on utils/build-script:
```diff
@@ -353,6 +353,21 @@ class BuildScriptInvocation(object):
+        # options.
+        if args.build_swift_sdk_overlay:
+            args.build_swift_dynamic_sdk_overlay = True
+            args.build_swift_static_sdk_overlay = False
```
Same here.

modocache commented:
@gribozavr OK, I addressed your comments, thanks to @rintaro's suggestion. Please have another look! I'd love to merge this before the weekend. 😎 ☀️ 

@swift-ci please test

gribozavr commented:
@modocache Thank you, my tests pass now!  Please merge when the CI succeeds!

modocache commented:
Thanks a ton, @gribozavr! 🙇
